File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/［PCCE 기출문제］ 10번 ／ 데이터 분석.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int sort_idx = -1;

bool cmp(vector<int> a, vector<int> b) {
    return a[sort_idx] < b[sort_idx];
}

vector<vector<int>> solution(vector<vector<int>> data, string ext, int val_ext, string sort_by) {
    vector<vector<int>> answer;
    
    // ext와 sort_by에 해당하는 인덱스를 찾습니다.
    vector<string> op = {"code", "date", "maximum", "remain"};
    int ext_idx = find(op.begin(), op.end(), ext) - op.begin();
    sort_idx = find(op.begin(), op.end(), sort_by) - op.begin();
    
    // 조건을 만족하는 데이터를 필터링합니다.
    for (int i = 0; i < data.size(); i++) {
        if (data[i][ext_idx] < val_ext) {
            answer.push_back(data[i]);
        }
    }
    
    // 필터링된 데이터를 정렬합니다.
    sort(answer.begin(), answer.end(), cmp);
    
    return answer;
}

File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/［PCCE 기출문제］ 10번 ／ 데이터 분석.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int[][] solution(int[][] data, String ext, int val_ext, String sort_by) {
        
        return Arrays.stream(data)
            .filter(i -> getValue(i, ext) < val_ext)
            .sorted(Comparator.comparing((int[] i) -> getValue(i, sort_by)))
            .toArray(int[][]::new);
    }

    
    int getValue(int[] item, String key){
        switch(key) {
            case "code" : return item[0];
            case "date" : return item[1];
            case "maximum" : return item[2];
            case "remain" : return item[3];
            default : return -1;
        }
    }
    
}